cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DocumentStore.java

```diff
@@ -0,0 +1,192 @@
+
+    /* This map should only be accessed inside synchronized(openDatastores) blocks */
+    private static final Map<File, DocumentStore> documentStores = new HashMap<File, DocumentStore>();
+
+    private static EventBus eventBus = new EventBus();
+
+    private static final Logger logger = Logger.getLogger(DocumentStore.class.getCanonicalName());
+
+    private DocumentStore(File location, KeyProvider keyProvider) throws DatastoreException, IOException, SQLException {
+        this.location = location;
+        this.databaseName = location.getName();
+        this.database = new DatabaseImpl(location, keyProvider);
+        this.query = new IndexManagerImpl(database);
+    }
+
+    /**
+     * <p>
+     * Get an instance of an existing or newly created store.
+     * </p>
+     * <p>
+     * Equivalent to calling {@link #getInstance(File, KeyProvider)} getInstance} with a
+     * {@link NullKeyProvider}.
+     * </p>
+     * @param location The location on the file system where the underlying files should be stored.
+     *                 Must be a directory.
+     * @return An existing or newly created store.
+     * @throws DatastoreNotCreatedException if the database cannot be opened
+     */
+    public static DocumentStore getInstance(File location) throws DatastoreNotCreatedException {
+        return getInstance(location, new NullKeyProvider());
+    }
+
+    /**
+     * <p>
+     * Get an instance of an existing or newly created store.
+     * </p>
+     * <p>
+     * If encryption is enabled for this platform, passing a {@link KeyProvider} containing a non-null
+     * key will open or create an encrypted database. If opening a database, the key used to
+     * create the database must be used. If encryption is not enabled for this platform, returning
+     * a non-null key will result in an exception.
+     * </p>
+     * @param location The location on the file system where the underlying files should be stored.
+     *                 Must be a directory.
+     * @param provider KeyProvider object. Use a {@link NullKeyProvider} if the database shouldn't
+     *                 be encrypted.
+     * @return An existing or newly created store.
+     * @throws DatastoreNotCreatedException if the database cannot be opened
+     */
+    public static DocumentStore getInstance(File location, KeyProvider provider) throws DatastoreNotCreatedException {
+        try {
+            synchronized (documentStores) {
+                DocumentStore ds = documentStores.get(location);
+                boolean created = false;
+                if (ds == null) {
+                    created = checkPathAndCreateIfNeeded(location);
+                    ds = new DocumentStore(location, provider);
+                    documentStores.put(location, ds);
+                    eventBus.post(new DatabaseOpened(ds.databaseName));
+                }
+                if (created) {
+                    eventBus.post(new DatabaseCreated(ds.databaseName));
```

Comment:
I think it's weird that this event would be posted after the `DatabaseOpened` in the creation case. Should probably move up.